compiler/rustc_middle/src/query/erase.rs

```diff
@@ -0,0 +1,234 @@
+use crate::mir;
+use crate::traits;
+use crate::ty;
+use rustc_ast::expand::allocator::AllocatorKind;
+use rustc_hir::MaybeOwner;
+pub use rustc_middle::traits::query::type_op;
+use std::intrinsics::type_name;
+use std::{
+    fmt,
+    mem::{size_of, transmute_copy, MaybeUninit},
+};
+
+#[derive(Copy, Clone)]
+pub struct Erased<T: Copy> {
+    data: MaybeUninit<T>,
+    #[cfg(debug_assertions)]
+    type_id: &'static str,
+}
+
+impl<T: Copy> fmt::Debug for Erased<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Erased")
+    }
+}
+
+pub trait EraseType: Copy {
+    type Result: Copy;
+}
+
+pub type Erase<T> = Erased<<T as EraseType>::Result>;
+
+#[inline(always)]
+pub fn erase<T: EraseType>(src: T) -> Erase<T> {
+    assert_eq!(
+        size_of::<T>(),
+        size_of::<T::Result>(),
+        "size of {} must match erased type {}",
+        type_name::<T>(),
+        type_name::<T::Result>()
+    );
+    Erased {
+        // SAFETY:: Is it safe to transmute to MaybeUninit
+        data: unsafe { transmute_copy(&src) },
+        #[cfg(debug_assertions)]
+        type_id: type_name::<T>(),
+    }
+}
+
+/// Restores an erased value.
+///
+/// This is only safe if `value` is a valid instance of `T`.
+/// For example if `T` was erased with `erase` previously.
+#[inline(always)]
+pub unsafe fn restore<T: EraseType>(value: Erase<T>) -> T {
+    #[cfg(debug_assertions)]
+    assert_eq!(value.type_id, type_name::<T>());
+    unsafe { transmute_copy(&value.data) }
+}
+
+impl<T> EraseType for &'_ T {
+    type Result = [u8; size_of::<*const ()>()];
+}
+
+impl<T> EraseType for &'_ [T] {
+    type Result = [u8; size_of::<*const [()]>()];
+}
+
+impl<T> EraseType for &'_ ty::List<T> {
+    type Result = [u8; size_of::<*const ()>()];
+}
```

Comment:
Why not `size_of::<Self>()`? Does it not compile because compiler thinks the size may be dependent on `T`?